### PR TITLE
[server] Guarantee to run HostContextProvider.ensureInitialized just …

### PR DESCRIPTION
**Pull request by geropl**
…once

## Related Issue(s)

Fixes #

## Release Notes

```release-note
NONE
```

### CODE REVIEW

File: components/server/src/auth/host-context-provider-impl.ts [hunk 1]
```diff
@@ -32,33 +34,41 @@ export class HostContextProviderImpl implements HostContextProvider {
...
+        const updateDynamicHostsJob = repeat(async () => {
+            const span = TraceContext.startSpan("updateDynamicHosts");
+            try {
+                await this.updateDynamicHosts({span});
+            } catch (error) {
+                log.error(`Failed to update dynamic hosts.`, error);
+            } finally {
+                span.finish();
+            }
+        }, 10000);
```
Comment (geropl) [inline components/server/src/auth/host-context-provider-impl.ts:48]: @AlexTugarev What's the use-case again where we have to sync this often?

And: can we change the "add host provider" code to:
 - add entry to the DB
 - PLUS execute the "addDynamicHostProvider" from below
so we don't rely on this timeout anymore?

Just saw the calls in honeycomg on staging, and I'd really love to set this to 30s or so, even with the optimization.

File: components/server/src/auth/host-context-provider-impl.ts [hunk 2]
```diff
@@ -32,33 +34,41 @@ export class HostContextProviderImpl implements HostContextProvider {
...
         }
-        this.createFixedHosts();
-
         try {
-            await this.updateDynamicHosts({ });
-        } catch (error) {
-            log.error(`Failed to update dynamic hosts.`, error);
-        }
+            this.initializationRunning = true;
+            this.createFixedHosts();
```
Comment (geropl) [inline components/server/src/auth/host-context-provider-impl.ts:61]: @AlexTugarev Note that we do not handle the error case here, but leave it in `this.initialized = false` state, keeping the current logic. Does that really make sense? :thinking:

File: components/server/src/auth/host-context-provider-impl.ts [hunk 3]
```diff
@@ -58,9 +53,17 @@ export class HostContextProviderImpl implements HostContextProvider {
                 span.finish();
             }
         }, 1999);
+        this.disposables.push(updateDynamicHostsJob);
+
         this.initialized = true;
```
Comment (AlexTugarev) [inline components/server/src/auth/host-context-provider-impl.ts:58]: let's moved this to first line of `init`. I think it's enough to signal a call to `init` was made. we don't need/want to wait for a first update.

Comment (geropl) [reply components/server/src/auth/host-context-provider-impl.ts:58]: I thought this should signal a defined state? If we do this we don't need `assertInitialized` at all, no? :thinking:

Comment (AlexTugarev) [reply components/server/src/auth/host-context-provider-impl.ts:58]: No. 

`assertInitialized` will throw if one would remove the `init` call by accident. Thus it needs to be checked in the public methods.

File: components/server/src/auth/host-context-provider-impl.ts [hunk 4]
```diff
@@ -58,9 +53,17 @@ export class HostContextProviderImpl implements HostContextProvider {
                 span.finish();
             }
         }, 1999);
+        this.disposables.push(updateDynamicHostsJob);
```
Comment (AlexTugarev) [inline components/server/src/auth/host-context-provider-impl.ts:56]: @geropl, wouldn't that already do the initial `await this.updateDynamicHosts` as well? 
So, we can remove the L.40 call, no?

**Pull request closed without merging**